e2etest/bgp_test.go

```diff
@@ -21,72 +21,107 @@ package e2e
 import (
 	"context"
 	"fmt"
-	"io/ioutil"
 	"net"
 	"strconv"
 	"time"
 
 	"go.universe.tf/metallb/e2etest/pkg/metrics"
 	"go.universe.tf/metallb/e2etest/pkg/routes"
 
+	"golang.org/x/sync/errgroup"
+
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/ginkgo/extensions/table"
 	. "github.com/onsi/gomega"
 	dto "github.com/prometheus/client_model/go"
-	"go.universe.tf/metallb/e2etest/pkg/executor"
 	"go.universe.tf/metallb/e2etest/pkg/frr"
 	frrconfig "go.universe.tf/metallb/e2etest/pkg/frr/config"
+	frrcontainer "go.universe.tf/metallb/e2etest/pkg/frr/container"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	clientset "k8s.io/client-go/kubernetes"
 	"k8s.io/kubernetes/test/e2e/framework"
 	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
 )
 
-const frrContainer = "frr"
+const (
+	frrIBGP = "frr-iBGP"
+	frrEBGP = "frr-eBGP"
+	IBGPAsn = 64512
+	EBGPAsn = 64513
+)
 
 var (
-	frrContainerIPv4 string
-	frrContainerIPv6 string
+	frrContainers []*frrcontainer.Frr
 )
 
+type containerConfig struct {
+	name string
+	nc   frrconfig.NeighborConfig
+	rc   frrconfig.RouterConfig
+}
+
 var _ = ginkgo.Describe("BGP", func() {
 	f := framework.NewDefaultFramework("bgp")
 	var cs clientset.Interface
-
+	containersConf := []containerConfig{
+		{
+			name: frrIBGP,
+			nc: frrconfig.NeighborConfig{
+				ASN: IBGPAsn,
+			},
+			rc: frrconfig.RouterConfig{
+				ASN:     IBGPAsn,
+				BGPPort: 179,
+			},
+		},
+		{
+			name: frrEBGP,
+			nc: frrconfig.NeighborConfig{
+				ASN: EBGPAsn,
+			},
+			rc: frrconfig.RouterConfig{
+				ASN:     IBGPAsn,
+				BGPPort: 180,
+			},
+		},
+	}
 	ginkgo.BeforeEach(func() {
 		cs = f.ClientSet
-
-		var err error
-		// If the FRR container is not running, start it on the local host.
-		if skipDockerCmd {
-			if len(frrTestConfigDir) == 0 {
-				framework.Fail("Missing FRR config directory.")
-			}
-		} else {
-			frrTestConfigDir, err = ioutil.TempDir("", "frr-conf")
-			framework.ExpectNoError(err)
-			err = frr.StartContainer(frrContainer, frrTestConfigDir)
-			framework.ExpectNoError(err)
+		frrContainers = make([]*frrcontainer.Frr, 0)
+		g := new(errgroup.Group)
+		for _, conf := range containersConf {
+			conf := conf
+			g.Go(func() error {
+				c, err := frrcontainer.New(conf.name, conf.nc, conf.rc, containersNetwork, hostContainerIPv4, hostContainerIPv6)
+				if c != nil {
```

Comment:
Because Start also calls updateVolumePermissions we could have a situation where a container is created (!=nil) but an error is returned. In that case we should add the container to the slice so it would be stopped.